module-system/module-implementations/sov-ibc/src/test_utils/sovereign/app.rs

```diff
@@ -0,0 +1,240 @@
+use ibc::core::ics04_channel::channel::{
+    ChannelEnd, Counterparty as ChanCounterparty, Order, State as ChannelState,
+};
+use ibc::core::ics04_channel::packet::Sequence;
+use ibc::core::ics04_channel::Version as ChannelVersion;
+use ibc::core::ics24_host::identifier::{ChainId, ChannelId, ClientId, ConnectionId, PortId};
+use ibc::core::ics24_host::path::{
+    ChannelEndPath, ClientConsensusStatePath, ClientStatePath, ConnectionPath, SeqAckPath,
+    SeqRecvPath, SeqSendPath,
+};
+use ibc::core::{ExecutionContext, ValidationContext};
+use ibc::Height;
+use sov_bank::Bank;
+use sov_modules_api::{Context, DaSpec};
+use sov_state::WorkingSet;
+use tendermint::{Hash, Time};
+
+use crate::context::clients::{AnyClientState, AnyConsensusState};
+use crate::context::IbcExecutionContext;
+use crate::test_utils::cosmos::helpers::dummy_tm_client_state;
+use crate::Ibc;
+
+/// Defines test fixture structure to interact with the bank and ibc modules
+pub struct TestApp<'a, C, Da>
+where
+    C: Context,
+    Da: DaSpec,
+{
+    chain_id: ChainId,
+    sdk_ctx: C,
+    ibc_ctx: IbcExecutionContext<'a, C, Da>,
+}
+
+impl<'a, C, Da> TestApp<'a, C, Da>
+where
+    C: Context,
+    Da: DaSpec + Clone,
+{
+    /// Initializes the test fixture
+    pub fn new(
+        chain_id: ChainId,
+        sdk_ctx: C,
+        ibc: &'a Ibc<C, Da>,
+        working_set: &'a mut WorkingSet<C::Storage>,
+    ) -> Self {
+        let shared_working_set = Rc::new(RefCell::new(working_set));
+
+        let ibc_execution_ctx = IbcExecutionContext {
+            ibc,
+            working_set: shared_working_set,
+        };
+
+        Self {
+            chain_id,
+            sdk_ctx,
+            ibc_ctx: ibc_execution_ctx,
+        }
+    }
+
+    pub fn chain_id(&self) -> &ChainId {
+        &self.chain_id
+    }
+
+    /// Returns access to the ibc module
+    pub fn working_set(&self) -> &Rc<RefCell<&'a mut WorkingSet<C::Storage>>> {
+        &self.ibc_ctx.working_set
+    }
+
+    pub fn sdk_ctx(&self) -> &C {
+        &self.sdk_ctx
+    }
+
+    /// Returns access to the bank module
+    pub fn bank(&self) -> &Bank<C> {
+        &self.transfer().bank
+    }
+
+    pub fn ibc_ctx(&self) -> IbcExecutionContext<'a, C, Da> {
+        self.ibc_ctx.clone()
+    }
+
+    /// Returns access to the transfer module
+    pub fn transfer(&self) -> &sov_ibc_transfer::Transfer<C> {
+        &self.ibc_ctx.ibc.transfer
+    }
+
+    pub fn get_balance_of(&self, user_address: C::Address, token_address: C::Address) -> u64 {
+        self.bank()
+            .get_balance_of(
+                user_address,
+                token_address,
+                &mut self.working_set().borrow_mut(),
+            )
+            .unwrap()
+    }
+
+    pub fn get_escrow_address(&self, token_address: C::Address) -> Option<C::Address> {
+        // Checks that the token has been transferred
+        self.transfer().escrowed_tokens.get(
+            &token_address.to_string(),
```

Comment:
This will only work for tokens that originate on the source chain (as we only use the stringified address as the key in the `escrowed_tokens` map [for these tokens](https://github.com/informalsystems/sovereign-sdk/blob/e0d552b17def36df690e0c3b4f95cd641df2c2cc/module-system/module-implementations/sov-ibc-transfer/src/call.rs#L87)).

Reply:
Right, I've updated comments and the naming. [b6b2f66](https://github.com/informalsystems/sovereign-sdk/pull/4/commits/b6b2f66eeaa6144694463e45c836d8d3fd47a8fd)
I hope it makes sense now